akka-http-core/src/main/scala/akka/http/impl/engine/http2/client/PersistentConnection.scala

```diff
@@ -57,18 +61,73 @@ private[http2] object PersistentConnection {
         override def onPull(): Unit =
           if (!hasBeenPulled(requestIn)) // requestIn might already have been pulled when we failed and went back to Unconnected
             pull(requestIn)
-
       }
 
       def connect(): Unit = {
         val requestOut = new SubSourceOutlet[HttpRequest]("PersistentConnection.requestOut")
         val responseIn = new SubSinkInlet[HttpResponse]("PersistentConnection.responseIn")
+        val connected = Source.fromGraph(requestOut.source)
+          .viaMat(connectionFlow)(Keep.right)
+          .toMat(responseIn.sink)(Keep.left)
+          .run()(subFusingMaterializer)
+
+        become(new Connecting(connected, requestOut, responseIn))
+      }
+
+      class Connecting(
+        connected:  Future[OutgoingConnection],
+        requestOut: SubSourceOutlet[HttpRequest],
+        responseIn: SubSinkInlet[HttpResponse]
+      ) extends State {
+        implicit val ec = materializer.executionContext
+        connected.onComplete {
+          case Success(_) =>
+            onConnected.invoke(())
+          case Failure(_) =>
+            onFailed.invoke(())
+        }
 
-        Source.fromGraph(requestOut.source)
-          .via(connectionFlow)
-          .runWith(responseIn.sink)(subFusingMaterializer)
+        var requestOutPulled = false
+        requestOut.setHandler(new OutHandler {
+          override def onPull(): Unit =
+            requestOutPulled = true
+          override def onDownstreamFinish(): Unit = ()
+        })
```

Comment:
We have already materialized the SubSourceOutlet, what does that mean for setting the handler after that is the pull lost perhaps?